External Content  FCREPO- 2748

bseeger opened the pull request:
**This PR revamps external content to meet the requirements in the Fedora API spec**
* * *

**JIRA Ticket**: https://jira.duraspace.org/browse/FCREPO-2748

* Other Relevant Links (Mailing list discussion, related pull requests, etc.)

# What does this Pull Request do?
Adds the foundation for External Content as described in the fedora spec. This implements external content redirect, proxy and copy. 

# What's new?
* This splits the lower fcrepo-kernel-modeshape binary classes out a little, to accommodate for URL redirects / proxies.  There are new classes like UrlBinary, LocalFileBinary and InternalFedoraBinary to handle dealing with different scenarios. 

* Copy happens in the fcrepo-http-api layer -- the modeshape level never even knows it's a copy.  You can find that code in FedoraLdp.java. 

# How should this be tested?
A description of what steps someone could take to:
*  Pull down the PR to compile - and play with the resulting war file.   Interaction with External Content will now require a new link header as outlined in the spec. 

Here are three simple interactions to test:  

Redirect: 
```
# create the resource
curl -X PUT -H "Link: <https://duraspace.org/>; rel=\"http://fedora.info/definitions/fcrepo#ExternalContent\"; handling=\"redirect\"; type=\"text/html\"" http://localhost:8080/rest/duraspace-homepage
# verify that the response code is a redirect with the link URI in the  Location header
curl -X GET http://localhost:8080/rest/duraspace-homepage
```
Proxy: 
```
# create the resource 
curl -v -X PUT -H "Link: <https://duraspace.org/wp-content/themes/duraspace/assets/images/whitedura.png>; rel=\"http://fedora.info/definitions/fcrepo#ExternalContent\"; handling=\"proxy\"; type=\"image/png\"" http://localhost:8080/rest/duraspace-logo.png

#verify that the resource was created and can be retrieved
curl -v -X GET http://localhost:8080/rest/duraspace-logo.png
```
Copy: 
```
# create resource 
curl -X PUT -H "Link: <http://localhost:8080/rest/duraspace-log.png>; rel=\"http://fedora.info/definitions/fcrepo#ExternalContent\"; handling=\"copy\"; type=\"image/png\"" http://localhost:8080/rest/duraspace-logo-copy.png

#verify that the resource was created: 
curl -v -X GET http://localhost:8080/rest/duraspace-logo-copy.png
```


Example:
* Does this change require documentation to be updated?   YES 

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo4/committers
@bbpennel @whikloj @dbernstein

bseeger commented:
To be honest, I haven't done any manual testing with this (there are tests using redirect/proxy -- but need more for copy) -- so it would be good to poke around with that a little.

whikloj commented on fcrepo-kernel-modeshape/src/test/java/org/fcrepo/kernel/modeshape/UrlBinaryTest.java:
```diff
@@ -0,0 +1,285 @@
+                        .withBody(EXPECTED_CONTENT)));
+        fileUrl = "http://localhost:" + wireMockRule.port() + "/file.txt";
+
+        mimeType = "text/plain";
```
Can this not be defined on line 77?

whikloj commented on fcrepo-kernel-modeshape/src/test/java/org/fcrepo/kernel/modeshape/UrlBinaryTest.java:
```diff
@@ -0,0 +1,285 @@
+        mockProxyProperty();
+
+        testObj.setContent(mockStream, mimeType, singleton(
+                new URI(checksum)), null, null);
```
There doesn't seem to be any verification in this test.

bseeger replied:
Fixed this in on of my later commits  - though it's not showing up here.

whikloj commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/ExternalContentHandler.java:
```diff
@@ -0,0 +1,227 @@
+
+        if (externalContentLinks.size() > 1) {
+            // got a problem, you can only have one ExternalContent links
+            // todo - throw error with constrained by info
```
Should this be added to the included constraint document and then that used here?

bseeger replied:
The spec says:  `Fedora servers must reject with a 4xx range status code requests that include multiple rel="http://fedora.info/definitions/fcrepo#ExternalContent" links.`    So, 400 works for me.

bseeger commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/FedoraExternalContent.java:
```diff
@@ -0,0 +1,31 @@
+ * @author bseeger
+ */
+
+public interface FedoraExternalContent {
```
I needed these somewhere where they could be accessed by the fcrepo-kernel-modeshape layer as well as the http layer... I could put them in the RdfLexicon, but wasn't sure that was quite right.  Or FedoraTypes?

bseeger commented on fcrepo-kernel-modeshape/src/test/java/org/fcrepo/integration/kernel/modeshape/LocalFileBinaryIT.java:
```diff
@@ -0,0 +1,219 @@
+    @Ignore
+    // Something in the ExternalResourceCacheEntry is returning
+    // binary (80B, SHA1=dcd59309841b261c56d45cfa6a587e46991c43c8)
+    // instead of a URI.
```
I'll run this w/o the Ignore - A lot has changed since your change, so perhaps this is fixed.

bbpennel replied:
I haven't looked at it, but let me know if you'd like me to @bseeger

bseeger replied:
I have not addressed this one yet and need to look at it some more. We can put in a ticket to fix this one after this PR goes through, maybe?

bseeger replied:
Created a ticket for this one so it doesn't get lost and we can go back and fix it: https://jira.duraspace.org/browse/FCREPO-2796

dbernstein commented:
@bseeger :  great great work @bseeger.  I added some sample code for testing above.  I was able to successfully perform a manual smoke test for proxy, redirect, and copy.

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/InternalFedoraBinary.java:
```diff
@@ -0,0 +1,406 @@
+    public RdfStream getFixity(final IdentifierConverter<Resource, FedoraResource> idTranslator,
+            final URI digestUri,
+            final long size) {
+        LOGGER.info("getFixity()");
```
Remove

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/InternalFedoraBinary.java:
```diff
@@ -0,0 +1,406 @@
+
+        try (final Timer.Context context = timer.time()) {
+
+            LOGGER.debug("Checking resource: " + getPath());
```
Not your change, but might as well switch this over to `LOGGER.debug("Checking resource: {}", getPath());`

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/InternalFedoraBinary.java:
```diff
@@ -0,0 +1,406 @@
+
+        try (final Timer.Context context = timer.time()) {
+
+            LOGGER.debug("Checking resource: " + getPath());
```
`LOGGER.debug("Checking resource: {}", getPath());`

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/LocalFileBinary.java:
```diff
@@ -0,0 +1,150 @@
+     * java.util.Collection, java.lang.String, org.fcrepo.kernel.api.services.policy.StoragePolicyDecisionPoint)
+     */
+    @Override
+    public void setContent(final InputStream content, final String contentType,
```
setContent should never be called on LocalFileBinary, only setExternalContent matters.  My suggestion would be to delete this and defer to `URLBinary`'s setContent method.

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/LocalFileBinary.java:
```diff
@@ -0,0 +1,150 @@
+        }
+    }
+
+    private static void decorateContentNode(final Node dsNode, final Node descNode, final Collection<URI> checksums,
```
This method can be deleted since it is only used by setContent

bseeger commented:
Thanks @dbernstein for adding those curl test commands and trying it out.

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/LocalFileBinary.java:
```diff
@@ -0,0 +1,150 @@
+            return sizeValue;
+        }
+        final File file = new File(getResourceUri().getPath());
+        setContentSize(file.length());
```
Pretty sure I wrote this, but this is a bad idea, we don't want to be changing a property of the description in a getter. A more appropriate place to retrieve the file size and set it would be in setExternalContent (if its not a memento). Maybe override setExternalContent, have it call super.setExternalContent, then call setContentSize.

bseeger replied:
I tried to change this and it needs more work.  Or rather, I tried to take this line out and do this instead: `setContentSize(getContentSize());` inside of setExternalContent.  It resulted in tests failing in a weird way. I can make a ticket to come back to this because I really want to get this PR in and then change things from there. Sound okay?

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/UrlBinary.java:
```diff
@@ -0,0 +1,343 @@
+        }
+        try {
+            final String content = IOUtils.toString(getResourceUri().toURL().openStream());
+            setContentSize(content.length());
```
Remove

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/UrlBinary.java:
```diff
@@ -0,0 +1,343 @@
+     * java.util.Collection, java.lang.String, org.fcrepo.kernel.api.services.policy.StoragePolicyDecisionPoint)
+     */
+    @Override
+    public void setContent(final InputStream content, final String contentType,
```
In my opinion, setContent for URLBinary's should throw a `UnsupportedOperationException` with a message about using setExternalContent instead. The relevant parts of this method (mainly decorating properties) should be moved up to setExternalContent.  We dont' need to create a jcr:data anymore, so that can be dropped.

bseeger replied:
Okay. I'm in agreement - I'll try this out to see what ripple effect it has... :)

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/UrlBinary.java:
```diff
@@ -0,0 +1,343 @@
+                fixityResults =
+                    CacheEntryFactory.forProperty(getProperty(REDIRECTS_TO)).checkFixity(algorithm);
+            } else {
+                LOGGER.warn("URL Binary -- not proxy or redirect, so what is it?");
```
👍 If you've made it into a URLBinary that is neither a proxy or a redirect, you've got problems. RepositoryRuntimeException with a little info is probably fine

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/UrlBinary.java:
```diff
@@ -0,0 +1,343 @@
+        try (final Timer.Context context = timer.time()) {
+
+            final String resourceLocation = getResourceLocation();
+            LOGGER.info("Checking external resource: " + resourceLocation);
```
switch to using {}

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/UrlBinary.java:
```diff
@@ -0,0 +1,343 @@
+                list = CacheEntryFactory.forProperty(getProperty(PROXY_FOR)).checkFixity(algorithms);
+            } else if (isRedirect()) {
+                list = CacheEntryFactory.forProperty(getProperty(REDIRECTS_TO)).checkFixity(algorithms);
+            }
```
else throw RepositoryRuntimeException

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/utils/ExternalResourceCacheEntry.java:
```diff
@@ -36,7 +36,7 @@
  * @since 2017-09-18
  */
 public class ExternalResourceCacheEntry extends BinaryCacheEntry {
```
Not sure if its worth pursuing right now, but I think we could probably remove the CacheEntryFactory at this point since it is exclusively being used for fixity checks inside two FedoraBinary implementations, which now have the internal knowledge about whether they are external or not.

bbpennel commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/utils/impl/CacheEntryFactory.java:
```diff
@@ -47,15 +53,21 @@ private CacheEntryFactory() {
+        // if it's an external binary, catch that here and treat it differently.
+        LOGGER.info("Did you know that property.getName() is: '{}' and Proxy: {} ", property.getName(),
+                PROXY_FOR.getNameSpace() + ":" + PROXY_FOR.getLocalName());
+        if (property.getName().endsWith(PROXY_FOR.getLocalName()) ||
```
Since this is checking the JCR property, it will have a name like `fedora:redirectsTo`, right? I think you can switch to:
`if (property.getName().equals(FedoraTypes.PROXY_FOR))` and the same with REDIRECTS_TO rather than using the RDFLexicon version of the same properties.

bbpennel commented on fcrepo-kernel-modeshape/src/test/java/org/fcrepo/integration/kernel/modeshape/FedoraBinaryImplIT.java:
```diff
@@ -524,6 +524,4 @@ public void testModifyDatastreamDescriptionLastMod() throws RepositoryException,
         }
     }
-
-
```
Revert change

bbpennel commented on fcrepo-kernel-modeshape/src/test/java/org/fcrepo/integration/kernel/modeshape/UrlBinaryIT.java:
```diff
@@ -0,0 +1,224 @@
+
+    @Test
+    public void testDatastream() throws Exception {
+        logger.info("Running testDatastream");
```
Can probably remove these info statements.

bbpennel commented on fcrepo-kernel-modeshape/src/test/java/org/fcrepo/kernel/modeshape/UrlBinaryTest.java:
```diff
@@ -0,0 +1,285 @@
+    }
+
+    @Test
+    public void testSetContent() throws Exception {
```
I don't think we need any setContent tests for URLBinary, other than maybe to verify that it throws an exception (see my comments in URLBinary)

Also, there aren't any tests for setExternalContent currently. I don't care too much if those get added or not since setExternalContent gets tested in ITs, so I'll leave if it up to you (or others) whether that's an issue.

bseeger replied:
Hi Ben, I just added a few that explicitly call setExternalContent, since I realized that I hadn't added any specific tests for that (though the setContent was covered quite well).

bbpennel commented on fcrepo-kernel-modeshape/src/test/java/org/fcrepo/kernel/modeshape/services/BinaryServiceImplTest.java:
```diff
@@ -133,4 +138,20 @@ public void testFindInvalidType() throws Exception {
     }
+
+    @Test
+    public void testFindOrCreateLocalFileBinary() throws Exception {
```
I think this test can go away, BinaryService doesn't know anything about external content (I think it did in an earlier version). So this file can probably be reset to master.

bbpennel commented on fcrepo-kernel-modeshape/src/test/java/org/fcrepo/kernel/modeshape/utils/TestHelpers.java:
```diff
@@ -40,6 +40,7 @@
 import javax.jcr.Value;
 
 import org.fcrepo.kernel.api.utils.ContentDigest;
+
```
Undo this change

bseeger commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/UrlBinary.java:
```diff
@@ -0,0 +1,343 @@
+    }
+
+    @Override
+    public void setExternalContent(final InputStream content, final String contentType,
```
Okay - yeah, that would be cleaner.

bseeger commented on fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java:
```diff
@@ -934,17 +934,13 @@ public void testEnableVersioning() throws Exception {
         createDatastream(id, "ds", "content");
 
         final String versionsUri = binaryUri + "/" + FCR_VERSIONS;
-        assertEquals("fcr:versions must not exist before enabling versioning",
```
because the logic of GET throws a repository exception if the item is not versioned: https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraVersioning.java#L272

Removing them made me save myself from pulling out my hair.  I'll try putting them back in again and see if it behaves any better now that other changes have been made.

bseeger commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/ExternalContentHandler.java:
```diff
@@ -0,0 +1,227 @@
+import org.slf4j.Logger;
+
+/**
+ * @author bseeger
```
addressed in one of my later commits, just not showing up here for some reason.

bbpennel replied:
Sorry to be a pain in the butt, but the standard in the rest of the code base is for the class description to start the javadoc, and the `@author` to follow. There also shouldn't be an empty line between the javadoc and the class declaration.

awoods commented:
On today's [Fedora Tech call](https://wiki.duraspace.org/display/FF/2018-05-10+-+Fedora+Tech+Meeting), the sentiment was that this PR is good to go. However, I see that @bbpennel and @whikloj still indicate requests for change.

bseeger commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/ExternalContentHandler.java:
```diff
@@ -0,0 +1,227 @@
+
+        final Map<String, String> map = link.getParams();
+        handling = map.get(HANDLING);
+        type = map.get(CONTENT_TYPE);
```
Can you point out the other areas you're talking about?

bseeger replied:
Ohhhh... I see what you mean... addressing this now.

bseeger commented:
Almost all of these have been address except for one by @whikloj where it talks about HANDLING and TYPE.

bseeger commented on fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraLdpIT.java:
```diff
@@ -208,6 +209,26 @@
     private static DateTimeFormatter tripleFormat =
       DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(of("GMT"));
 
+    private String getExternalContentLinkHeader(final String url, final String handling, final String mimeType) {
```
Sure, I'll move it down.

bseeger commented on fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraLdpIT.java:
```diff
@@ -2041,7 +2050,7 @@ public void testGetObjectGraphVariants() throws IOException {
     @Test
     public void testGetObjectGraph() throws IOException {
-        logger.trace("Entering testGetObjectGraph()...");
+        LOGGER.trace("Entering testGetObjectGraph()...");
```
Yeah, this is confusing. There is `logger` from the parent object and `LOGGER` declared in this file (before I got here). I changed a few trying to figure out which should actually be there (since the pattern is that `LOGGER` is usually used).  I shouldn't have touched any of them and left it alone. Changing it back now. 

Someone should clean this up at some point, unless there's a good reason to have `logger` from parent and `LOGGER` in this class.

bseeger commented on fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraLdpIT.java:
```diff
@@ -3080,16 +3089,15 @@ public void testEmbeddedContainedResources() throws IOException {
 
     @Test
-    public void testExternalMessageBody() throws IOException {
+    public void testExternalMessageBodyRedirect() throws IOException {
```
Created this ticket to not lose this idea.   https://jira.duraspace.org/browse/FCREPO-2797

bseeger commented:
Given how Github keeps abbreviating diffs on me, I missed a few comments. Addressing those now.  I think Github is mad at me.

bseeger commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java:
```diff
@@ -697,6 +749,9 @@ public Response createObject(@HeaderParam(CONTENT_DISPOSITION) final ContentDisp
 
                 if (hasVersionedResourceLink(links)) {
                     resource.enableVersioning();
+                    if (resource instanceof FedoraBinary) {
```
Hmm... no I don't think it's necessary and I'll move it back. I think it got shuffled around when I was trying to get things to work again and I said "I'll put it here and move it later".  Guess it's later. ;)

bseeger commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraBinaryImpl.java:
```diff
@@ -491,7 +312,8 @@ private boolean hasDescriptionProperty(final String relPath) {
 
-    private Property getDescriptionProperty(final String relPath) {
+    @Override
+    protected Property getDescriptionProperty(final String relPath) {
```
Thanks for catching that!

bbpennel commented:
I don't think you will be able to call `getContentSize()` on the object
before setContentSize has been called, since its not set yet.
`setExternalContent` will likely need to be overridden for this class with
something like:
```

super.setExternalContent(...);

final File file = new File(getResourceUri().getPath());
setContentSize(file.length());

```

On Fri, May 18, 2018 at 4:08 PM, B. Seeger <notifications@github.com> wrote:

> *@bseeger* commented on this pull request.
> ------------------------------
>
> In fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/
> modeshape/LocalFileBinary.java
> <https://github.com/fcrepo4/fcrepo4/pull/1357#discussion_r189380518>:
>
> > +    public String getMimeType() {
> +        return getMimeTypeValue();
> +    }
> +
> +    /*
> +     * (non-Javadoc)
> +     * @see org.fcrepo.kernel.modeshape.FedoraBinaryImpl#getContentSize()
> +     */
> +    @Override
> +    public long getContentSize() {
> +        final long sizeValue = super.getContentSize();
> +        if (sizeValue > -1L) {
> +            return sizeValue;
> +        }
> +        final File file = new File(getResourceUri().getPath());
> +        setContentSize(file.length());
>
> I tried to change this and it needs more work. Or rather, I tried to take
> this line out and do this instead: setContentSize(getContentSize());
> inside of setExternalContent. It resulted in tests failing in a weird way.
> I can make a ticket to come back to this because I really want to get this
> PR in and then change things from there. Sound okay?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fcrepo4/fcrepo4/pull/1357#discussion_r189380518>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA7MUhUvC3buJrTVGAmD47lpyvjembdzks5tzyo3gaJpZM4T3Jaw>
> .
>

bseeger commented:
I've addressed all the comments and created a few tickets for things I'm not addressing right now.   This is the ticket to address moving the setContentSize() out of getContentSize() (addressing one of @bbpennel comments above).

bseeger commented:
I was thinking about setting it explicitly, but wasn't sure of the best path forward. Do you want to see that logic in this PR or one shortly after this one?

bbpennel commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraLdp.java:
```diff
@@ -697,6 +735,9 @@ public Response createObject(@HeaderParam(CONTENT_DISPOSITION) final ContentDisp
                 if (hasVersionedResourceLink(links)) {
                     resource.enableVersioning();
+                    if (resource instanceof FedoraBinary) {
+                        resource.getDescription().enableVersioning();
```
This should be `resource.enableVersioning();`, FedoraBinary has the special enableVersioning method that enables versioning on the description as well.

bbpennel commented on fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/exceptionhandlers/ExternalMessageBodyExceptionMapper.java:
```diff
@@ -0,0 +1,66 @@
+import javax.ws.rs.ext.Provider;
+
+/**
+ * If an External Content Link header isn't formatted correctly or missing something, return
```
I think this description is supposed to be swapped with the javadoc on `toResponse`

bbpennel commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BinaryService.java:
```diff
@@ -26,23 +26,21 @@
  * @since 10/10/14
  */
 public interface BinaryService extends Service<FedoraBinary> {
```
Looks like this file could still use a reversion to master, it appears that one extra space was added to the beginning of every line

bbpennel commented on fcrepo-kernel-modeshape/src/test/java/org/fcrepo/kernel/modeshape/services/BinaryServiceImplTest.java:
```diff
@@ -49,6 +49,8 @@
 @RunWith(MockitoJUnitRunner.class)
 public class BinaryServiceImplTest {
 
+    private static final String LOCAL_FILE_RESOURCE = "file:///path/to/file";
```
Not used, can be removed.

bseeger commented on fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/exceptionhandlers/ExternalMessageBodyExceptionMapper.java:
```diff
@@ -0,0 +1,61 @@
+    private static final Logger LOGGER = getLogger(ExternalMessageBodyExceptionMapper.class);
+
+    @Context
+    private UriInfo uriInfo;
```
I had meant to use a different base class -- switched that around and this is necessary now.

bseeger commented on fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/exceptionhandlers/ExternalMessageBodyExceptionMapper.java:
```diff
@@ -0,0 +1,61 @@
+    private UriInfo uriInfo;
+
+    @Context
+    private ServletContext context;
```
I had meant to use a different base class -- switched that around and this is necessary now.
